controlplane-rs/src/gateway_controller.rs

```diff
@@ -0,0 +1,178 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use futures::StreamExt;
+use std::{
+    ops::Sub,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use crate::*;
+use k8s_gateway_api::{Gateway, GatewayClass, GatewayClassSpec, GatewaySpec};
+use k8s_openapi::api::core::v1::Service;
+use kube::{
+    api::{Api, ListParams, Patch, PatchParams},
+    runtime::{controller::Action, watcher::Config, Controller},
+    Resource, ResourceExt,
+};
+
+use chrono::Utc;
+use gateway_utils::*;
+use k8s_openapi::apimachinery::pkg::apis::meta::v1 as metav1;
+use tracing::*;
+
+pub async fn reconcile(gateway: Arc<Gateway>, ctx: Arc<Context>) -> Result<Action> {
+    let start = Instant::now();
+    let client = ctx.client.clone();
+    let name = gateway.name_any();
+    let ns = gateway.namespace().unwrap();
```

Comment:
A flat `unwrap()` here might be problematic, if I'm not mistaken you can in fact have no namespace declared and that will mean the resource belongs to the default namespace right? If you find that's accurate, then we can probably do something like:

```suggestion
    let ns = gateway.namespace().unwrap_or("default".into());
```

Which would be safe. However we should ideally use an existing constant for `"default"` if one exists in kube-rs or somewhere nearby.
```